File: macos/Sources/abacad/ScreenshotCache.swift
import Foundation

// A short-lived screenshot cache with single-flight coalescing, the desktop
// analogue of the Android client's shot cache (AbacadAccessibilityService).
//
// Two independent optimizations, both keyed off a monotonic screen "generation":
//
//   • Cache — a capture is reused for up to CACHE_WINDOW_NS after it was taken,
//     so the dashboard's periodic poll and an agent's screenshot that land within
//     the same second share one capture + JPEG encode instead of doing two.
//
//   • Single-flight — concurrent screenshot commands (each dispatched on its own
//     detached Task) that miss the cache join one in-flight capture and share its
//     result, rather than each kicking an independent ScreenCaptureKit grab.
//
// Any drive command (tap, click, scroll, input, …) calls invalidate(), bumping
// the generation so the very next screenshot is guaranteed fresh — never a frame
// that predates the action. macOS has no platform screenshot rate limit (unlike
// Android's ~333ms takeScreenshot floor), so there is deliberately no pacing or
// retry logic here; the cache exists purely to avoid redundant work.
actor ScreenshotCache {
    static let shared = ScreenshotCache()

    /// A screenshot requested within this window of the last capture is served
    /// from cache instead of re-capturing.
    private static let CACHE_WINDOW_NS: UInt64 = 1_000_000_000 // 1s

    private struct Entry {
        let shot: ScreenCapture.Shot
        let tree: [String: Any]?
        let stampNs: UInt64
        let gen: Int
    }

    private var gen = 0                      // monotonic screen generation
    private var entry: Entry?                // last successful capture
    private var task: Task<Entry, Error>?    // in-flight single-flight capture
    private var taskGen = -1                 // generation the in-flight capture belongs to
    private var taskHasTree = false          // whether the in-flight capture will carry a UI tree
    private var taskId = 0                    // identity token (Task is a value type, so no ===)

    /// Invalidate the cache after a screen-changing command, so the next
    /// screenshot re-captures. Called for every non-screenshot method.
    func invalidate() {
        gen &+= 1
    }

    /// Serve a screenshot as the wire result dict, reusing a cached or in-flight
    /// capture when possible. `includeTree` requires the served frame to carry a
    /// UI tree; a cached frame without one (or an in-flight capture not taking
    /// one) does not satisfy a tree request.
    func screenshot(includeTree: Bool) async throws -> [String: Any] {
        let now = DispatchTime.now().uptimeNanoseconds
        if let e = entry, e.gen == gen,
           now &- e.stampNs < Self.CACHE_WINDOW_NS,
           !includeTree || e.tree != nil {
            return response(e, includeTree: includeTree)
        }
        let e = try await capture(includeTree: includeTree)
        return response(e, includeTree: includeTree)
    }

    /// Join a compatible in-flight capture, or kick a fresh one. Coalescing is
    /// safe under actor reentrancy: while the creator awaits the Task, later
    /// callers re-enter, see the same `task` for this generation, and await it too.
    private func capture(includeTree: Bool) async throws -> Entry {
        if let t = task, taskGen == gen, !includeTree || taskHasTree {
            return try await t.value
        }
        let capGen = gen
        let wantTree = includeTree
        let t = Task { () throws -> Entry in
            let shot = try await ScreenCapture.capture()
            // Capture the tree alongside the pixels so both describe the same
            // screen state; AccessibilityTree.capture() is synchronous.
            let tree = wantTree ? AccessibilityTree.capture() : nil
            return Entry(shot: shot, tree: tree,
                         stampNs: DispatchTime.now().uptimeNanoseconds, gen: capGen)
        }
        taskId &+= 1
        let myId = taskId
        task = t; taskGen = capGen; taskHasTree = wantTree
        do {
            let e = try await t.value
            // Clear only if no newer capture replaced ours — so a completed task
            // can't be joined as if still in flight, and we don't clobber a
            // successor. (Task is a value type; identity is via this token.)
            if taskId == myId { task = nil }
            // Only cache if no drive command bumped the generation mid-capture;
            // otherwise the frame predates that action and must not be reused.
            if e.gen == gen { entry = e }
            return e
        } catch {
            if taskId == myId { task = nil }
            throw error
        }
    }

    /// Build the wire result from a captured frame, attaching the UI tree only
    /// when asked (and present). Field stays `png_base64` for wire compatibility
    /// even though the bytes are JPEG — see ScreenCapture.
    private func response(_ e: Entry, includeTree: Bool) -> [String: Any] {
        var result: [String: Any] = ["w": e.shot.w, "h": e.shot.h, "png_base64": e.shot.base64]
        if includeTree, let tree = e.tree { result["tree"] = tree }
        return result
    }
}

File: windows/ScreenshotCache.cs
namespace Abacad;

// A short-lived screenshot cache with single-flight coalescing, the Windows port
// of macos/ScreenshotCache.swift. Two optimizations, both keyed off a monotonic
// screen "generation":
//
//   • Cache — a capture is reused for up to CacheWindowMs after it was taken, so
//     the dashboard's periodic poll and an agent's screenshot that land within the
//     same second share one capture + JPEG encode instead of doing two.
//
//   • Single-flight — concurrent screenshot commands that miss the cache join one
//     in-flight capture and share its result, rather than each kicking an
//     independent GDI grab.
//
// Any drive command (tap, click, scroll, input, …) calls Invalidate(), bumping the
// generation so the very next screenshot is guaranteed fresh — never a frame that
// predates the action.
sealed class ScreenshotCache
{
    public static readonly ScreenshotCache Shared = new();

    const long CacheWindowMs = 1000; // 1s

    sealed record Entry(ScreenCapture.Shot Shot, Dictionary<string, object?>? Tree, long StampMs, int Gen);

    readonly object _gate = new();
    int _gen;
    Entry? _entry;
    Task<Entry>? _inflight;
    int _inflightGen = -1;
    bool _inflightHasTree;
    int _taskId;

    /// Invalidate the cache after a screen-changing command, so the next screenshot
    /// re-captures. Called for every non-screenshot method.
    public void Invalidate()
    {
        lock (_gate) _gen++;
    }

    /// Serve a screenshot as the wire result dict, reusing a cached or in-flight
    /// capture when possible. `includeTree` requires the served frame to carry a UI
    /// tree; a cached frame without one does not satisfy a tree request.
    public async Task<Dictionary<string, object?>> Screenshot(bool includeTree)
    {
        Task<Entry> task;
        lock (_gate)
        {
            long now = Environment.TickCount64;
            if (_entry is { } e && e.Gen == _gen && now - e.StampMs < CacheWindowMs
                && (!includeTree || e.Tree != null))
                return Response(e, includeTree);

            if (_inflight != null && _inflightGen == _gen && (!includeTree || _inflightHasTree))
            {
                task = _inflight;
            }
            else
            {
                int capGen = _gen;
                bool wantTree = includeTree;
                int myId = ++_taskId;
                task = Capture(capGen, wantTree, myId);
                _inflight = task;
                _inflightGen = capGen;
                _inflightHasTree = wantTree;
            }
        }
        var result = await task.ConfigureAwait(false);
        return Response(result, includeTree);
    }

    async Task<Entry> Capture(int capGen, bool wantTree, int myId)
    {
        try
        {
            var shot = await Task.Run(() => ScreenCapture.Capture()).ConfigureAwait(false);
            // Capture the tree alongside the pixels so both describe the same screen
            // state.
            var tree = wantTree ? await Task.Run(() => UiTree.Capture()).ConfigureAwait(false) : null;
            var e = new Entry(shot, tree, Environment.TickCount64, capGen);
            lock (_gate)
            {
                if (_taskId == myId) _inflight = null;
                // Only cache if no drive command bumped the generation mid-capture;
                // otherwise the frame predates that action and must not be reused.
                if (e.Gen == _gen) _entry = e;
            }
            return e;
        }
        catch
        {
            lock (_gate) { if (_taskId == myId) _inflight = null; }
            throw;
        }
    }

    // Field stays `png_base64` for wire compatibility even though the bytes are JPEG.
    static Dictionary<string, object?> Response(Entry e, bool includeTree)
    {
        var result = new Dictionary<string, object?>
        {
            ["w"] = e.Shot.W,
            ["h"] = e.Shot.H,
            ["png_base64"] = e.Shot.Base64,
        };
        if (includeTree && e.Tree != null) result["tree"] = e.Tree;
        return result;
    }
}
